Write optimization in AbstractCsvWriter class

ncapponi opened the pull request:
This is the fix suggested for the following issue: https://github.com/super-csv/super-csv/issues/71

ZioberMichal commented on super-csv/src/main/java/org/supercsv/io/AbstractCsvWriter.java:
```diff
@@ -64,26 +63,26 @@ public AbstractCsvWriter(final Writer writer, final CsvPreference preference) {
 			throw new NullPointerException("preference should not be null");
 		}
-		
-		this.writer = new BufferedWriter(writer);
```
This change brakes backward compatibility. I think that we should add new constructor which takes additional constructor.

ZioberMichal commented:
Hi @ncapponi, great PR but could you please:
1. Join these two commits into one
2. You have committed a lot of changes related with empty lines. Could you revert them? It will be much simpler to analyse PR.

ncapponi commented:
Thanks for your answer. 
I just pushed a new version with a single commit and removed empty lines changes.
Also, I've introduced a new constructor to keep compatibility with existing code, as you asked for. 
This new constructor allow to skip the wrapping of the provided writer providing an additional boolean. Let me know if you want any other change.

ncapponi commented:
Note that I've changed the type of the *writer* field to a Writer instead of a BufferedWriter, but this has no consequence on the code, as my change only make use of the *write(String str)* method of Writer class.

ZioberMichal commented:
Now, when all changes are made in one commit and without "empty  lines" I see how good this PR actually is. :+1: 

One question, why you cast delimiter char to `char`? It should work as well without it.

ncapponi commented:
The cast is necessary because StringBuilder.append() method can accept both int and char arguments, and the delimiter is returned as an integer.

ncapponi commented:
I've pushed yet another version because I added a new constructor in AbstractCSVWriter class previously but forgot to add one in concrete classes. For now, I've just added the new constructor in CsvMapWriter.
Tell me if you want that I add it to all concrete classes.
Alternatively, I could provide static factory methods to make construction easier from the user point of view, e.g.

```
public static createCsvMapWriter() { // return a map writer with wrapped writer }
public static createUnwrappedCsvMapWriter() { // return a map writer with unwrapped writer }
```

ZioberMichal commented:
It looks great how it is now. @kbilsted what do you think about this factory? Will you merge this PR shortly? I can wait and release new version with it.

kbilsted commented:
I only see one commit, addbace0ceab759f3e3dce7a55e596ba60025f20

I'm still at a loss regarding why this is faster than using the BufferedWriter - the whole purpose of a buffered writer is to reduce the number of writes. Possibly even more than what this PR is doing. Anybody care to dig into the buffered implementation for an explanation?

laurentvaills commented:
As you can see in the source code, each time you call a `write` method from the `BufferedWriter` class a `synchronized` block is executed, which takes some time to acquire the lock. And currently supercsv calls the method `write` for very small parts of the row. Then it's more efficient to use a `StringBuilder` a calling only once the write method with the full row.
Wrapping the `Writer` with a `BufferedWriter` force to have these synchronized blocks, but sometimes you do not want to buffer at all, or you may want to tweak the buffer's size. In the current implementation, we can't tweak the size of the buffer of the `BufferedWriter`.

ncapponi commented:
No @kbilsted I've not committed the static methods. It is just an alternative proposal in case you don't like the commit that contains an additional constructor.

Concerning performance and usage of BufferedWriter, there is two things to consider:
- the usage of a StringBuilder to build the whole row instead of multiple calls to a Writer (BufferedWriter or not) is clearly faster because you have no synchronization. It is in effect like a small buffer without locking. Note that with the current implementation, there is at least 2*N synchronized sections where N is the number of cell per row.
- usage of a BufferedWriter is usually a good thing, but it should be up to the client to provide one. In our use case, we don't want one because we have an underlying custom writer that does some caching. Having said that, I agree that by default a BufferedWriter should be better for performance in a single threaded environment - but high number of locks implied by point 1 may completely remove this advantage. 

Concerning your remark on BufferedWriter implementation, I do think the buffering happens because Writer.write(String str) method delegates to Writer.write(String str, int off, int len) which is overridden in BufferedWriter.

I think the PR should satisfy you because by default all implementations will benefit from the StringBuilder improvement and BufferedWriter is still used. Only the clients wanting to use the provided Writer will skip the BufferedWriter.

ncapponi commented:
Sorry my message is partially redundant with laurentvaills' s one but he posted it while I was writing mine...

ncapponi commented:
I've done some performance tests again to give more detail on the gain provided by the PR, using the existing StandardCsvWritingBenchmarkTest test already implemented in super-csv (I didn't used it initially because it only tests a single-threaded scenario). See the related issue https://github.com/super-csv/super-csv/issues/71

kbilsted commented:
Let's get it in there! ;)

ncapponi commented:
Thanks @kbilsted !

kbilsted commented:
Thanks for all your work @ncapponi  would you be interested in helping out with the issues? I'd prefer to be standing on the side line for a bit.

ncapponi commented:
@kbilsted Sorry I'm not interested in being involved in the project. I hope you'll find people to help you.
